### DISCUSSION TITLE
Can't Pull Docker Image:

### DISCUSSION THREAD

**Discussion by dschinkel** [community/community#26023] [Actions]
<p>I can’t figure how how to get the second script right, after I create the image:</p>
<p>I’m trying to move away from using a script I had been using to build and push docker images where I was using <a href="https://github.com/docker/build-push-action" rel="nofollow noopener">docker/build-push-action</a>.</p>
<p>I don’t care about retaining the old images from GitHub Packages Docker registry.  I just want to push a new image to the Google Container Registry at this point.</p>
<p>I am now running a script to create the image myself and push it to the new <strong>Google Container Registry</strong> route.  So here are my new jobs:</p>
<pre><code class="lang-auto">  build_docker_image_from_artifact:
    needs: build
    name: Build a docker image based off the build artifact
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/download-artifact@v2
        with:
          name: dist-artifact
          path: dist
      - run: docker build . --file Dockerfile --tag dschinkel/we-do-tdd:latest
</code></pre>
<p><img src="https://user-images.githubusercontent.com/338798/181094831-31d69248-5c21-4509-9ea3-c992c6b56008.jpeg" alt="Screen Shot 2020-09-01 at 11.56.37 PM" data-base62-sha1="2dNEwgWjnTLCwpezrilEY2dIMx3" width="300" height="33"></p>
<p>Now <strong>I’m not really sure what the image name is that it actually creates</strong>.  I see that it generated a sha or hash or something and my tag.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b495f07da36771d66d127366cd2cbda6001e5976.jpeg" data-download-href="/uploads/short-url/pLxbGd5kBr9V6xWVpYcLEMV6VvM.jpeg?dl=1" title="Screen Shot 2020-09-02 at 12.26.31 AM" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/338798/181094832-7fc1d237-69d0-4fcc-a9aa-0969973d8d44.jpeg" alt="Screen Shot 2020-09-02 at 12.26.31 AM" data-base62-sha1="pLxbGd5kBr9V6xWVpYcLEMV6VvM" width="690" height="51" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b495f07da36771d66d127366cd2cbda6001e5976_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-09-02 at 12.26.31 AM</span><span class="informations">700×52 15 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Apparently <strong><code>push ghcr.io/dschinkel/we-do-tdd:latest</code></strong> isn’t quite right though in my next script when trying to reference it:</p>
<pre><code class="lang-auto">  push_image_to_github_container_registry:
    needs: login_github_container_registry
    name: Push Docker image to GitHub Container Repository.  Creates a Docker Image Package
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/download-artifact@v2
        with:
          name: dist-artifact
          path: dist
      - run: ls dist
      # ghcr.io/OWNER/IMAGE_NAME.latest
      - run: docker push ghcr.io/dschinkel/we-do-tdd:latest
</code></pre>
<p>I get:</p>
<pre><code class="lang-auto">Run docker push ghcr.io/dschinkel/we-do-tdd:latest
An image does not exist locally with the tag: ghcr.io/dschinkel/we-do-tdd
The push refers to repository [ghcr.io/dschinkel/we-do-tdd]
</code></pre>
<p><strong>Other Info</strong> (in case it helps)<br>
My last push that was using my old script based on <a href="https://github.com/docker/build-push-action" rel="nofollow noopener">docker/build-push-action</a>that I built a couple days ago generated this image but not sure what my new script is actually creating because I haven’t packaged it yet:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/73d774cca4dbbe9e10b513bf00fa1ec77350f7eb.jpeg" data-download-href="/uploads/short-url/gwMwmvdArcgR48HZW0ho5DnkJ8n.jpeg?dl=1" title="Screen Shot 2020-09-02 at 12.01.03 AM" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/338798/181094833-084b1788-66f7-452c-812e-e419805be86a.jpeg" alt="Screen Shot 2020-09-02 at 12.01.03 AM" data-base62-sha1="gwMwmvdArcgR48HZW0ho5DnkJ8n" width="690" height="369" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/73d774cca4dbbe9e10b513bf00fa1ec77350f7eb_2_690x369.jpeg, https://aws1.discourse-cdn.com/github/original/2X/7/73d774cca4dbbe9e10b513bf00fa1ec77350f7eb.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/73d774cca4dbbe9e10b513bf00fa1ec77350f7eb.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/73d774cca4dbbe9e10b513bf00fa1ec77350f7eb_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-09-02 at 12.01.03 AM</span><span class="informations">800×428 102 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I took a look at <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#example-of-updated-workflow" rel="nofollow noopener">Migrating to GitHub Container Registry for Docker images</a> but not sure if I’m missing something related to that or it’s just a dumb mistake I made above somewhere that I’m just not seeing.  I want to think it’s maybe the latter…?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/dschinkel">@dschinkel</a> ,</p>
<p>Before you push an image to a Docker or container registry, you need to build (if need) and tag the image.</p>
<pre><code class="lang-auto">docker build -t REPOSITORY:TAG -f Dockerfile .
</code></pre>
<p>OR</p>
<pre><code class="lang-auto">docker tag IMAGE_ID REPOSITORY:TAG
</code></pre>
<p>The <strong>REPOSITORY</strong> should contain the registry’s domain and name of the target repository or owner on the registry, so that the ‘<strong>docker push</strong>’ command can know where the image is pushed to.</p>
<p>If the <strong>REPOSITORY</strong> does not contain registry’s information when you tag the image, and only set the registry’s information when pushing the image, the ‘<strong>docker push</strong>’ command will not recognize registry’s information and it will return the error message like as you have seen:</p>
<pre><code class="lang-auto">Run docker push ghcr.io/dschinkel/we-do-tdd:latest
An image does not exist locally with the tag: ghcr.io/dschinkel/we-do-tdd
The push refers to repository [ghcr.io/dschinkel/we-do-tdd]
</code></pre>
<p>This also has been mentioned in “<a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images" rel="nofollow noopener"><strong>Migrating to GitHub Container Registry for Docker images</strong></a>”.</p>
<blockquote>
<p><a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#migrating-a-docker-image-using-the-docker-cli" rel="nofollow noopener">Re-tag the image with the new domain and a new image name.</a></p>
</blockquote>

**Comment by dschinkel** [comment]
<p>Yes I’m already building it, above I have:</p>
<pre><code class="lang-auto">run: docker build . --file Dockerfile --tag dschinkel/we-do-tdd:latest
</code></pre>
<p>I’ve since then gone a little further.</p>
<p>Pretty sure I need to be using the Image ID not we-do-tdd in my <strong>push</strong> job.  So I’m working on trying to persist that newly created image ID as an environment variable that my push job can use:</p>
<pre><code class="lang-auto">  build_docker_image_from_artifact:
    needs: build
    name: Build Docker Image Off Artifact
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/download-artifact@v2
        with:
          name: dist-artifact
          path: dist
      - run: docker build . --file Dockerfile --tag latest
      - run: echo "::set-env name=IMAGE_NAME::$(docker images latest --format "{{.ID}}")"
</code></pre>
<p>Notice this updated script sets it as an environment variable called <code>IMAGE_NAME</code> that holds the newly created image ID.  Problem I have now is it’s not persisting that environment variable to my next job:</p>
<pre><code class="lang-auto">  push_image_to_github_container_registry:
    needs: login_github_container_registry
    name: Push image to GitHub Container Repository
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      # ghcr.io/OWNER/IMAGE_NAME.latest
      # - run: docker push ghcr.io/dschinkel/${{ env.IMAGE_NAME }}.latest
      - run: docker push ghcr.io/dschinkel/${IMAGE_NAME}.latest
</code></pre>
<p>Problem here now is <code>${IMAGE_NAME}.latest</code> - <code>${IMAGE_NAME}</code> isn’t being persisted from the environment variable I set in the previous job.</p>
<p>I have verified that this does indeed set the environment variable:<br>
<code>run: echo "::set-env name=IMAGE_NAME::$(docker images latest --format "{{.ID}}")"</code></p>
<p>but just my last job when it tries to access it, that value is lost.</p>
<p>And again I’m not <strong>MOVING</strong> an <strong>EXISTING</strong> image, I’m creating a <strong>NEW</strong> image and on the fly, trying to push that up to the new GitHub Container Repository.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/dschinkel">@dschinkel</a> ,</p>
<p>In a job, the environment variable you set via the ‘<strong>set-env</strong>’ command in a step will be only available to the subsequent steps in the same job. The steps in other jobs can’t access this environment variable. More details, see “<a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="nofollow noopener">Setting an environment variable</a>”.</p>
<p>If you want to access this value in the subsequent job, you can set it as an output of the previous job.<br>
In your case, you can do like as this:</p>
<pre><code class="lang-yaml">jobs:
  build_docker_image_from_artifact:
    outputs:
      IMAGE_NAME: ${{ steps.get-image-name.outputs.IMAGE_NAME }}
    . . .
    steps:
      . . .
      - name: get image name
        id: get-image-name
        run: echo "::set-output name=IMAGE_NAME::$(docker images latest --format "{{.ID}}")"

  push_image_to_github_container_registry:
    needs: [build_docker_image_from_artifact, . . .]
    . . .
    steps:
      . . .
      - name: push image
        run: docker push ghcr.io/dschinkel/${{ needs.build_docker_image_from_artifact.outputs.IMAGE_NAME }}.latest
</code></pre>

**Comment by dschinkel** [comment]
<p>Thanks very much! I’ll try this, I was spending hours on trying to figure out this combination!  I’m off to bed so will try it when I wake for the day</p>

**Comment by BrightRan** [comment]
<p>you’re welcome, <a class="mention" href="/dschinkel">@dschinkel</a> .<br>
Any progress, feel free to tell me.</p>
<p>In addition, the following is an example mentioned in the docs about workflow syntax, you also can reference to it:<br>
<a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" class="onebox" target="_blank" rel="nofollow noopener">https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs</a></p>

**Comment by dschinkel** [comment]
<p>So I was able to get the environment variable working, thanks for the tip.  I hadn’t seen that part of the documentation for some reason.</p>
